tools/chplcheck/src/rules.py

```diff
@@ -17,93 +17,136 @@
+            context, node
+        )
 
     @driver.basic_rule(Record)
-    def CamelCaseRecords(context, node):
+    def CamelCaseRecords(context: Context, node: Record):
         """
         Warn for records that are not 'camelCase'.
         """
 
-        return check_camel_case(node)
+        return check_camel_case(context, node)
 
     @driver.basic_rule(Function)
-    def CamelCaseFunctions(context, node):
+    def CamelCaseFunctions(context: Context, node: Function):
         """
         Warn for functions that are not 'camelCase'.
         """
 
         # Override functions / methods can't control the name, that's up
         # to the parent.
-        if node.is_override(): return True
+        if node.is_override():
+            return True
 
-        if node.linkage() == 'extern': return True
-        if node.kind() == 'operator': return True
-        if node.name() == 'init=': return True
+        if node.linkage() == "extern":
+            return True
+        if node.kind() == "operator":
+            return True
+        if node.name() == "init=":
+            return True
 
-        return check_camel_case(node)
+        return check_camel_case(context, node)
 
     @driver.basic_rule(Class)
-    def PascalCaseClasses(context, node):
+    def PascalCaseClasses(context: Context, node: Class):
         """
         Warn for classes that are not 'PascalCase'.
         """
 
-        return check_pascal_case(node)
+        return check_pascal_case(context, node)
 
     @driver.basic_rule(Module)
-    def PascalCaseModules(context, node):
+    def PascalCaseModules(context: Context, node: Module):
         """
         Warn for modules that are not 'PascalCase'.
         """
 
-        return node.kind() == "implicit" or check_pascal_case(node)
+        return node.kind() == "implicit" or check_pascal_case(context, node)
 
     @driver.basic_rule(Module, default=False)
-    def UseExplicitModules(context, node):
+    def UseExplicitModules(_, node: Module):
         """
         Warn for code that relies on auto-inserted implicit modules.
         """
 
         return node.kind() != "implicit"
 
     @driver.basic_rule(Loop)
-    def DoKeywordAndBlock(context, node):
+    @driver.basic_rule(SimpleBlockLike)
+    def DoKeywordAndBlock(
+        context: Context, node: typing.Union[Loop, SimpleBlockLike]
+    ):
         """
         Warn for redundant 'do' keyword before a curly brace '{'.
         """
 
-        return node.block_style() != "unnecessary"
+        check = node.block_style() != "unnecessary"
+        if not check:
+            lines = context.get_file_text(node.location().path()).split("\n")
+            text = range_to_text(node.location(), lines)
+            # TODO: this should be smarter about the do keyword
+            text = re.sub(r"\bdo( *)", "", text, 1)
+            # remove any trailing whitespace
+            text = re.sub(r" +\n", "\n", text)
```

Comment:
The trick I described with clamping etc. can work here too to avoid regular expressions.